src/chronify/time_series_checker.py

```diff
@@ -24,54 +24,95 @@ def __init__(self, conn: Connection, table: Table, schema: TableSchema) -> None:
+        count = self._check_expected_timestamps()
+        self._check_null_consistency()
+        self._check_expected_timestamps_by_time_array(count)
 
-    def _check_expected_timestamps(self) -> None:
+    def _check_expected_timestamps(self) -> int:
         expected = self._time_generator.list_timestamps()
         time_columns = self._time_generator.list_time_columns()
         stmt = select(*(self._table.c[x] for x in time_columns)).distinct()
         for col in time_columns:
             stmt = stmt.where(self._table.c[col].is_not(None))
         df = read_database(stmt, self._conn, self._schema.time_config)
         actual = self._time_generator.list_distinct_timestamps_from_dataframe(df)
+
+        if isinstance(self._schema.time_config, DatetimeRange):
+            expected = [pd.Timestamp(x) for x in expected]
         check_timestamp_lists(actual, expected)
+        return len(expected)
+
+    def _check_null_consistency(self) -> None:
+        # If any time column has a NULL, all time columns must have a NULL.
+        time_columns = self._time_generator.list_time_columns()
+        if len(time_columns) == 1:
+            return
 
-    def _check_expected_timestamps_by_time_array(self) -> None:
-        tmp_name = make_temp_view_name()
-        self._run_timestamp_checks_on_tmp_table(tmp_name)
-        self._conn.execute(text(f"DROP TABLE IF EXISTS {tmp_name}"))
+        all_are_null = " AND ".join((f"{x} IS NULL" for x in time_columns))
+        any_are_null = " OR ".join((f"{x} IS NULL" for x in time_columns))
+        query_all = f"SELECT COUNT(*) FROM {self._schema.name} WHERE {all_are_null}"
+        query_any = f"SELECT COUNT(*) FROM {self._schema.name} WHERE {any_are_null}"
+        res_all = self._conn.execute(text(query_all)).fetchone()
+        assert res_all is not None
+        res_any = self._conn.execute(text(query_any)).fetchone()
+        assert res_any is not None
+        if res_all[0] != res_any[0]:
+            msg = (
+                "If any time columns have a NULL value for a row, all time columns in that "
+                "row must be NULL. "
+                f"Row count where all time values are NULL: {res_all[0]}. "
+                f"Row count where any time values are NULL: {res_any[0]}. "
+            )
+            raise InvalidTable(msg)
 
-    def _run_timestamp_checks_on_tmp_table(self, table_name: str) -> None:
+    def _check_expected_timestamps_by_time_array(self, count: int) -> None:
         id_cols = ",".join(self._schema.time_array_id_columns)
-        filters = [f"{x} IS NOT NULL" for x in self._time_generator.list_time_columns()]
-        where_clause = " AND ".join(filters)
+        time_cols = ",".join(self._schema.time_config.list_time_columns())
+        # NULL consistency was checked above.
+        where_clause = f"{self._time_generator.list_time_columns()[0]} IS NOT NULL"
+        on_expr = " AND ".join([f"t1.{x} = t2.{x}" for x in self._schema.time_array_id_columns])
```

Comment:
For simplicity, `in id_cols` for on_expr and t1_id_cols

Reply:
But `id_cols` is a string, not a list.